File: ios/App/App/RPCClient.swift
//
//  ContentView.swift
//  App
//
//  Created by Slugalisk on 8/20/20.
//  Copyright © 2020 MemeLabs. All rights reserved.
//

import Foundation
import SwiftProtobuf
import PromiseKit

//enum RPCClientError: Error {
//    case runtimeError(String)
//}

struct RPCClientError: Error {
    let message: String

    init(_ message: String) {
        self.message = message
    }

    public var localizedDescription: String {
        return message
    }
}

struct RPCEvent: OptionSet {
    let rawValue: Int
    
    static let data    = RPCEvent(rawValue: 1 << 0)
    static let close    = RPCEvent(rawValue: 1 << 1)
    static let requestError  = RPCEvent(rawValue: 1 << 2)
    static let responseError  = RPCEvent(rawValue: 1 << 3)
}

class RPCResponseStream<T: Message> {
    public let close: () -> Void;
    public var delegate: (T?, RPCEvent) -> Void = {_, _ in}
    
    public init(_ close: @escaping () -> Void) {
        self.close = close
    }
}

extension OutputStream {
    var data: Data? {
        return self.property(forKey: .dataWrittenToMemoryStreamKey) as? Data
    }
}

class RPCClient {
    private static let callbackMethod = "_CALLBACK"
    private static let cancelMethod   = "_CANCEL"
    private static let anyURLPrefix = "strims.gg/"
    
    private var nextCallID: UInt64 = 0;
    private var callbacks: Dictionary<UInt64, (PBCall) -> Void> = Dictionary()
    private var g: SwiftBridge
    
    init() {
        self.g = SwiftBridge()
        self.g.onData = {(_ b: Data?) -> Void in
            self.handleCallback(b)
        }
    }
    
    private func handleCallback(_ b: Data?) {
        let stream = InputStream(data: b!)
        stream.open()
        
        do {
            let call = try BinaryDelimited.parse(messageType: PBCall.self, from: stream)
            
            let json = try call.jsonString()
            print("received rpc data: \(json)")
            
            let callback = self.callbacks[call.parentID]
            callback?(call)
        } catch {
            print("error: \(error)")
        }
        
        stream.close()
    }
    
    private func getNextCallID() -> UInt64 {
        self.nextCallID += 1
        return self.nextCallID
    }
    
    private func call<T: Message>(_ method: String, _ arg: T, _ callID: UInt64, _ parentID: UInt64 = 0) throws {
        let arg = try Google_Protobuf_Any(message: arg, typePrefix: RPCClient.anyURLPrefix)
        let call = PBCall.with {
            $0.id = callID
            $0.parentID = parentID
            $0.method = method
            $0.argument = arg
        }
        
        let stream = OutputStream.toMemory()
        stream.open()
        try BinaryDelimited.serialize(message: call, to: stream)
        stream.close()
        
        self.g.write(stream.data!)
    }
    
    public func call<T: Message>(_ method: String, _ arg: T) {
        DispatchQueue.global(qos: .default).async {
            do {
                try self.call(method, arg, self.getNextCallID())
            } catch {}
        }
    }
    
    public func callStreaming<T: Message, R: Message>(_ method: String, _ arg: T) -> RPCResponseStream<R> {
        let callID = self.getNextCallID()
    
        let stream = RPCResponseStream<R>({
            self.callbacks.removeValue(forKey: callID)
            do {
                try self.call(RPCClient.cancelMethod, PBCancel(), self.getNextCallID(), callID)
            } catch {}
        })
        
        self.callbacks[callID] = {(_ call: PBCall) -> Void in
            do {
                let messageType = Google_Protobuf_Any.messageType(forTypeURL: call.argument.typeURL)
                switch (messageType) {
                case is R.Type:
                    let arg = try R(serializedData: call.argument.value)
                    stream.delegate(arg, RPCEvent.data)
                case is PBClose.Type:
                    self.callbacks.removeValue(forKey: callID)
                    stream.delegate(nil, RPCEvent.close)
                case is PBError.Type:
                    fallthrough
                default:
                    self.callbacks.removeValue(forKey: callID)
                    stream.delegate(nil, RPCEvent.responseError)
                }
            } catch {
                stream.delegate(nil, RPCEvent.responseError)
            }
        }
        
        DispatchQueue.global(qos: .default).async {
            do {
                try self.call(method, arg, callID)
            } catch {
                self.callbacks.removeValue(forKey: callID)
                stream.delegate(nil, RPCEvent.requestError)
            }
        }
        
        return stream
    }

    public func callUnary<T: Message, R: Message>(_ method: String, _ arg: T) -> Promise<R> {
        Promise { seal in
            let callID = self.getNextCallID()

            let timer = Timer.scheduledTimer(withTimeInterval: 5.0, repeats: false) { timer in
                self.callbacks.removeValue(forKey: callID)
                seal.reject(RPCClientError("call timeout"))
            }
            
            self.callbacks[callID] = {(_ call: PBCall) -> Void in
                self.callbacks.removeValue(forKey: callID)
                timer.invalidate()
                                
                do {
                    let messageType = Google_Protobuf_Any.messageType(forTypeURL: call.argument.typeURL)
                    switch (messageType) {
                    case is R.Type:
                        let arg = try R(serializedData: call.argument.value)
                        seal.resolve(arg, nil)
                    case is PBError.Type:
                        let error = try PBError(serializedData: call.argument.value)
                        seal.reject(RPCClientError(error.message))
                    default:
                        print("unexpected response type: \(call.argument.typeURL)")
                        seal.reject(RPCClientError("unexpected response type: \(call.argument.typeURL)"))
                    }
                } catch {
                    seal.reject(error)
                }
            }
            
            DispatchQueue.global(qos: .default).async {
                do {
                    try self.call(method, arg, callID)
                } catch {
                    self.callbacks.removeValue(forKey: callID)
                    seal.reject(error)
                }
            }
        }
    }
}

File: android/app/src/main/java/gg/strims/ppspp/RPCClient.kt
package gg.strims.ppspp

import android.util.Log
import com.google.protobuf.Message
import gg.strims.ppspp.proto.Rpc
import java.io.ByteArrayOutputStream
import java.util.*
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.Future
import java.util.concurrent.Semaphore
import kotlin.concurrent.timerTask

class RPCClientError(message: String) : Exception(message)

data class RPCEvent(val rawValue: Int) : BitSet() {
    companion object {
        val data: RPCEvent = RPCEvent(rawValue = 1 shl 0)
        val close: RPCEvent = RPCEvent(rawValue = 1 shl 1)
        val requestError: RPCEvent = RPCEvent(rawValue = 1 shl 2)
        val responseError: RPCEvent = RPCEvent(rawValue = 1 shl 3)
    }
}

class RPCResponseStream<T : Message>(val close: () -> Unit) {
    var delegate: (T?, RPCEvent) -> Unit = { _: T?, _: RPCEvent -> }
}

open class RPCClient(filepath: String) {
    companion object {
        const val TAG = "RPCClient"

        // const val callbackMethod = "_CALLBACK"
        const val cancelMethod = "_CANCEL"
        const val anyURLPrefix = "strims.gg/"
    }

    val executor: ExecutorService = Executors.newCachedThreadPool()

    private var nextCallID: Long = 0
    var callbacks: MutableMap<Long, (Rpc.Call) -> Unit> = mutableMapOf()
    private var g: AndroidBridge = AndroidBridge(filepath)

    init {
        this.g.onData = { b: ByteArray? -> this.handleCallback(b) }
    }

    private fun handleCallback(b: ByteArray?) {
        val stream = b!!.inputStream()
        try {
            // parse call
            val call = Rpc.Call.parseDelimitedFrom(stream)
            this.callbacks[call.parentId]?.let {
                it(call)
            } ?: throw RPCClientError("missing callback")
        } catch (e: Exception) {
            Log.e(TAG, "could not parse message", e)
        }
    }

    fun getNextCallID(): Long {
        this.nextCallID += 1
        return this.nextCallID
    }

    fun <T : Message> call(method: String, arg: T, callID: Long, parentID: Long = 0) {
        val packed = com.google.protobuf.Any.pack(arg, anyURLPrefix)

        val call = Rpc.Call.newBuilder()
            .setParentId(parentID)
            .setId(callID)
            .setArgument(packed)
            .setMethod(method)
            .build()

        val stream = ByteArrayOutputStream()
        call.writeDelimitedTo(stream)
        stream.close()

        this.g.write(stream.toByteArray())
    }

    fun <T : Message> call(method: String, arg: T) {
        this.call(method, arg, this.getNextCallID())
    }

    // inline required to get refined type information, see https://stackoverflow.com/a/46870546/5698680
    inline fun <T : Message, reified R : Message> callStreaming(
        method: String,
        arg: T
    ): RPCResponseStream<R> {
        val callID = this.getNextCallID()
        val stream = RPCResponseStream<R> {
            this.callbacks.remove(callID)
            try {
                this.call(
                    cancelMethod,
                    Rpc.Cancel.getDefaultInstance(),
                    this.getNextCallID(),
                    callID
                )
            } catch (e: Exception) {
                Log.e(TAG, "failed to send call", e)
            }
        }

        this.callbacks[callID] = {
            try {
                when {
                    it.argument.`is`(R::class.java) -> {
                        val unpacked = it.argument.unpack(R::class.java)
                        stream.delegate(unpacked, RPCEvent.data)
                    }
                    it.argument.`is`(Rpc.Close::class.java) -> {
                        callbacks.remove(callID)
                        stream.delegate(null, RPCEvent.close)
                    }
                    it.argument.`is`(Rpc.Error::class.java) -> {
                        callbacks.remove(callID)
                        stream.delegate(null, RPCEvent.responseError)
                    }
                    else -> { // error
                        callbacks.remove(callID)
                        stream.delegate(null, RPCEvent.responseError)
                    }
                }
            } catch (e: Exception) {
                stream.delegate(null, RPCEvent.responseError)
            }
        }

        try {
            this.call(method, arg, callID)
        } catch (e: Exception) {
            this.callbacks.remove(callID)
            stream.delegate(null, RPCEvent.requestError)
            throw e
        }

        return stream
    }

    inline fun <T : Message, reified R : Message> callUnary(method: String, arg: T): Future<R> {
        val callId = getNextCallID()
        return executor.submit<R> {
            // set timeout
            val timer = Timer()
            val s = Semaphore(1)
            var ex: Throwable? = null
            timer.schedule(timerTask {
                callbacks.remove(callId)
                ex = RPCClientError("call timeout")
                s.release()
            }, 5L * 1000) // five seconds


            // prepare callback
            var result: R? = null
            s.acquire()
            callbacks[callId] = {
                callbacks.remove(callId)
                timer.cancel()
                when {
                    it.argument.`is`(R::class.java) -> {
                        result = it.argument.unpack(R::class.java)
                        Log.i(TAG, result.toString())
                    }
                    it.argument.`is`(Rpc.Error::class.java) -> {
                        val error = it.argument.unpack(Rpc.Error::class.java)
                        ex = RPCClientError(error.message)
                    }
                    else -> {
                        ex = RPCClientError("unexpected response type ${it.argument.typeUrl}")
                    }
                }
                s.release()
            }

            // call method
            try {
                call(method, arg, callId)
            } catch (e: Exception) {
                callbacks.remove(callId)
                throw e
            }

            // wait for response
            s.acquire()
            if (ex != null) {
                throw ex as Throwable
            }
            return@submit result
        }
    }
}
